gapic-generator-cloud/templates/cloud/client/_resource.erb

```diff
@@ -0,0 +1,28 @@
+<%- assert_locals resource -%>
+##
+# Create a fully-qualified <%= resource.name %> resource string.
+#
+# The resource will be in the following format:
+#
+# `<%= resource.template %>`
+#
+<%- (resource.positional_args + resource.named_args).each do |name, _regexp, msg| -%>
+# @param <%= name %> [String]
+<%- if msg -%>
+#   <%= msg %>
+<%- end -%>
+<%- end -%>
+#
+# @return [String]
+<%- args = resource.positional_args.map(&:first) -%>
+<%- args += resource.named_args.map(&:first).map { |a| "#{a}:"} -%>
```

Comment:
I'm a bit torn on this, I think having the named parameters be keyword arguments makes increases readability of usages of these methods but we might have to keep these all as positional args. Changing a path template from `thing/*` to `thing/{thing_id=*}` will become a breaking change and I think that might be surprising for an API producer.

Reply:
The way it is implemented now, `thing/*` generates a positional arg, and `thing/{foo}` generates a named arg `foo`. I can't think of any instance where a resource path is using positional templates, though.

Where positional templates do show up is in named fields, so something like `thing/{foo=bar/*/baz/**}`. In this case, the path method would have a `foo` named arg and ensure that the value passes matched the `bar/*baz/**` pattern.

Reply:
I see. Thanks for being so thorough! I was confused and spitballed through this with @lukesneeringer. I found out that the design of the patterns that are passed through the `google.api.resource` will disallows template patterns with wild cards. I think the wild cards in that python resource handler is also accounting for paths in `google.api.http` which we won't be generating path helpers for.

The patterns will always look like: `projects/{project}` where `{project}` can be treated as `.+`. This would then simplify this implementation a lot in which we simply create keyword args for those named parts and simply check if they are non-empty strings.

Reply:
Further more we don't have to worry about positional arguments because they will always be named as part of the `google.api.resource` spec.

Reply:
That was my thinking. But, the functionality was there from the python code so I figured it wouldn't do any harm to allow for that usage, even if it isn't very likely.

Reply:
Gotcha. Let's only support the named params in this PR. I'd like to reduce the number of things we support to only things that are explicitly supported by the spec.

Reply:
So you want the generator to raise if a resource has positional arguments?

Reply:
To be more explicit, the following situations are invalid for a `google.api.resource.pattern`.

1. a pattern that contains a wildcard `projects/*/topics/*`
2. a pattern that contains an id bound to a pattern. `projects/{project="*"}`

Let's raise on both of these situations.

Reply:
Both are invalid? K.

Reply:
Both are invalid for `google.api.resource.pattern` but are valid for `google.api.http.path`. Since we are deriving our data from `google.api.resource` we don't need to account for these cases.

Reply:
Updated. PTAL.